File: project/app/src/main/java/org/owntracks/android/data/repos/MemoryContactsRepo.java
package org.owntracks.android.data.repos;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.owntracks.android.injection.scopes.PerApplication;
import org.owntracks.android.model.FusedContact;
import org.owntracks.android.model.messages.MessageCard;
import org.owntracks.android.model.messages.MessageLocation;
import org.owntracks.android.support.ContactImageProvider;
import org.owntracks.android.support.Events;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import timber.log.Timber;

@PerApplication
public class MemoryContactsRepo implements ContactsRepo {
    public static final long MAJOR_STEP = 1000000;

    private final EventBus eventBus;
    private final ContactImageProvider contactImageProvider;

    private final MutableLiveData<Map<String, FusedContact>> contacts = new MutableLiveData<>(new HashMap<>());

    private long majorRevision = 0;
    private long revision = 0;

    @Inject
    public MemoryContactsRepo(EventBus eventBus, ContactImageProvider contactImageProvider) {
        this.eventBus = eventBus;
        this.contactImageProvider = contactImageProvider;
        this.eventBus.register(this);
    }

    @Override
    public MutableLiveData<Map<String, FusedContact>> getAll() {
        return contacts;
    }

    @Override
    public FusedContact getById(String id) {
        return contacts.getValue().get(id);
    }


    private synchronized void put(String id, final FusedContact contact) {
        Timber.v("new contact allocated id:%s, tid:%s", id, contact.getTrackerId());
        Map<String, FusedContact> map = contacts.getValue();
        map.put(id,contact);
        contacts.postValue(map);
    }

    @Override
    @MainThread
    public synchronized void clearAll() {
        contacts.getValue().clear();

        majorRevision -= MAJOR_STEP;
        revision = 0;
        contactImageProvider.invalidateCache();
    }

    @Override
    public synchronized void remove(String id) {
        Timber.v("removing contact: %s", id);
        final FusedContact c = contacts.getValue().remove(id);

        if (c != null) {
            c.setDeleted();
            eventBus.post(new Events.FusedContactRemoved(c));

            majorRevision -= MAJOR_STEP;
            revision = 0;
        }
    }

    @Override
    public void update(@NonNull String id, @NonNull MessageCard m) {
        FusedContact c = getById(id);

        if (c != null) {
            c.setMessageCard(m);
            contactImageProvider.invalidateCacheLevelCard(c.getId());

            revision++;
            eventBus.post(c);

        } else {
            c = new FusedContact(id);
            c.setMessageCard(m);
            contactImageProvider.invalidateCacheLevelCard(c.getId());

            put(id, c);
            revision++;
            eventBus.post(new Events.FusedContactAdded(c));
        }
    }

    @Override
    public synchronized void update(@NonNull String id, @NonNull MessageLocation messageLocation) {
        FusedContact fusedContact = getById(id);

        if (fusedContact != null) {
            // If timestamp of last location message is <= the new location message, skip update. We either received an old or already known message.
            if (fusedContact.setMessageLocation(messageLocation)) {
                revision++;
                eventBus.post(fusedContact);
            }
        } else {
            fusedContact = new FusedContact(id);
            fusedContact.setMessageLocation(messageLocation);
            put(id, fusedContact);
            revision++;
            eventBus.post(new Events.FusedContactAdded(fusedContact));
        }
    }

    // Allows for quickly checking if items in the repo were updated/added (getRevison > savedRevision) or removed (getRevision < savedRevision)
    public long getRevision() {
        return majorRevision + revision;
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onEventMainThread(Events.ModeChanged e) {
        clearAll();
    }


    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onEventMainThread(Events.EndpointChanged e) {
        clearAll();
    }
}

File: project/app/src/main/java/org/owntracks/android/data/repos/MemoryContactsRepo.kt
package org.owntracks.android.data.repos

import javax.inject.Inject
import javax.inject.Singleton
import kotlin.time.measureTime
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.owntracks.android.model.Contact
import org.owntracks.android.model.messages.MessageCard
import org.owntracks.android.model.messages.MessageLocation
import org.owntracks.android.model.messages.MessageTransition
import org.owntracks.android.support.ContactBitmapAndName
import org.owntracks.android.support.ContactBitmapAndNameMemoryCache
import timber.log.Timber

@Singleton
class MemoryContactsRepo
@Inject
constructor(
    private val contactBitmapAndNameMemoryCache: ContactBitmapAndNameMemoryCache,
) : ContactsRepo {

  private val contacts = mutableMapOf<String, Contact>()
  private val repoLock = Mutex()

  private val mutableRepoChangedEvent = MutableSharedFlow<ContactsRepoChange>()

  override val repoChangedEvent: SharedFlow<ContactsRepoChange> = mutableRepoChangedEvent
  override val all: Map<String, Contact>
    get() = contacts

  override fun getById(id: String): Contact? {
    return contacts[id]
  }

  override suspend fun clearAll() {
    Timber.i("Clearing all contacts. Waiting for lock")
    measureTime {
          repoLock.withLock {
            Timber.v("Lock acquired")
            contacts.clear()
            contactBitmapAndNameMemoryCache.evictAll()
            mutableRepoChangedEvent.emit(ContactsRepoChange.AllCleared)
          }
        }
        .also { Timber.d("Cleared all contacts in $it") }
  }

  override suspend fun remove(id: String) {
    Timber.v("removing contact: $id. waiting for lock")
    measureTime {
          repoLock.withLock {
            Timber.v("Lock acquired")
            contacts.remove(id)?.run {
              mutableRepoChangedEvent.emit(ContactsRepoChange.ContactRemoved(this))
            }
          }
        }
        .also { Timber.d("remove contact $id took $it") }
  }

  override suspend fun update(id: String, messageCard: MessageCard) {
    measureTime {
          Timber.v("updating contact card for contact $id. waiting for repoLock")
          repoLock.withLock {
            Timber.v("Lock acquired")

            getById(id)?.apply {
              // We just received new contact details, so invalidate the cache entry.
              contactBitmapAndNameMemoryCache.remove(id)
              this.setMessageCard(messageCard)
              mutableRepoChangedEvent.emit(ContactsRepoChange.ContactCardUpdated(this))
            } ?: run { Contact(id).apply { setMessageCard(messageCard) }.also { put(id, it) } }
          }
        }
        .also { Timber.d("update contact card for contact $id took $it") }
  }

  override suspend fun update(id: String, messageLocation: MessageLocation) {
    Timber.v("Updating location for contact $id from $messageLocation. waiting for repoLock")
    updateContactLocation(id) { contact: Contact ->
      contact.setLocationFromMessageLocation(messageLocation)
    }
  }

  override suspend fun update(id: String, messageTransition: MessageTransition) {
    Timber.v("Updating location for contact $id from $messageTransition. waiting for repoLock")
    updateContactLocation(id) { contact: Contact ->
      contact.setLocationFromMessageTransition(messageTransition)
    }
  }

  private suspend fun updateContactLocation(id: String, updateLocation: (Contact) -> Boolean) {

    measureTime {
          repoLock.withLock {
            getById(id)?.apply {
              // If timestamp of last location message is <= the new location message, skip update.
              // We either received an old or already known message.
              if (updateLocation(this)) {
                mutableRepoChangedEvent.emit(ContactsRepoChange.ContactLocationUpdated(this))
              }
            }
                ?: run { // If getById is null, we have not seen this contact id before
                  Contact(id)
                      .apply {
                        updateLocation(this)
                        // We may have seen this contact id before, and it may have been removed
                        // from the repo Check the cache to see if we have a name
                        contactBitmapAndNameMemoryCache[id]?.also {
                          if (it is ContactBitmapAndName.CardBitmap && it.name != null) {
                            setMessageCard(MessageCard().apply { name = it.name })
                          }
                        }
                      }
                      .also { put(id, it) }
                }
          }
        }
        .also { Timber.d("update location for contact $id took $it") }
  }

  private suspend fun put(id: String, contact: Contact) {
    Timber.v("new contact allocated id=$id, tid=${contact.trackerId}")
    contacts[id] = contact
    mutableRepoChangedEvent.emit(ContactsRepoChange.ContactAdded(contact))
  }
}
